src/mpid/ch4/shm/xpmem/xpmem_control.h

```diff
@@ -99,15 +79,52 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_XPMEM_ctrl_send_lmt_req_cb(MPIDI_SHM_ctrl_hdr
 
     if (rreq) {
         /* Matching receive was posted */
-        MPIR_Comm_release(root_comm);   /* -1 for posted_list */
         MPIDIG_REQUEST(rreq, rank) = slmt_req_hdr->src_rank;
         MPIDIG_REQUEST(rreq, tag) = slmt_req_hdr->tag;
         MPIDIG_REQUEST(rreq, context_id) = slmt_req_hdr->context_id;
 
         /* Complete XPMEM receive */
-        mpi_errno = MPIDI_XPMEM_handle_lmt_recv(slmt_req_hdr->src_offset, slmt_req_hdr->data_sz,
-                                                slmt_req_hdr->sreq_ptr, slmt_req_hdr->src_lrank,
-                                                rreq);
+#ifdef MPIDI_CH4_SHM_XPMEM_COOP_P2P
+        int recvtype_iscontig;
+        MPIDI_av_entry_t *av;
+        /* TODO: need to support non-contig datatype on receiver side. */
+        MPIR_Datatype_iscontig(MPIDIG_REQUEST(rreq, datatype), &recvtype_iscontig);
+        if (recvtype_iscontig) {
+            MPIDI_XPMEM_REQUEST(rreq, counter) =
+                &MPIDI_XPMEM_global.coop_counter[MPIDI_XPMEM_global.local_rank *
+                                                 MPIDI_XPMEM_global.num_local +
+                                                 slmt_req_hdr->src_lrank];
```

Comment:
I assume by the way this data is laid out that there can only be one PXMEM data transfer going on at a time, right?

Reply:
There can be only one data transfer at a time between any pair of the sender and receiver. The following transmission is valid (can happen at the same time):
rank 0 sends to rank 1
rank 1 sends to rank 0

Reply:
:+1: